### PR TITLE
API example clients

### PR DESCRIPTION
**Pull request by ihorserba**
Error handling not yet complete

### CODE REVIEW

**Comment by ihorserba** [comment]
TODO: need to call stop_zello_stream() once interrupted by user using ctrl+C

**Comment by ihorserba** [comment]
> TODO: need to call stop_zello_stream() once interrupted by user using ctrl+C

Added such logic with global variables. Not pretty but the simplest way.

**Comment by Skowt** [comment]
Wanted to call out that I got this working using Django + Celery to run a background task using your example. Happy to discuss it if you want. It doesn't have any code added around interruptions but it looks like Celery is handling that fine.

**Comment by ihorserba** [comment]
Need to calculate delay between packet sending more precisely.
Another idea - send the existing media with smaller delays, calculate the remaining time and send empty audio frames during this time.

**Comment by ihorserba** [comment]
JS client code has been reviewed by @megamk in https://github.com/zelloptt/zello-channel-api/pull/112.
Pushed the appropriate changes here.

File: examples/stream.cs [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+    using System.Threading;
+    using System.Text.Json;
+    using System.Buffers.Binary;
+    using Microsoft.Extensions.Configuration;
+
+    // https://tools.ietf.org/html/rfc7845
+    // https://tools.ietf.org/html/rfc3533
+    class OpusFileStream : IDisposable {
+
+        Stream opusFile;
```
Comment (vocoded) [inline examples/stream.cs:16]: properties should follow PascalCase convention

File: examples/stream.cs [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+    using System.Collections.Generic;
+    using System.Net.WebSockets;
+    using System.Threading;
+    using System.Text.Json;
+    using System.Buffers.Binary;
+    using Microsoft.Extensions.Configuration;
+
+    // https://tools.ietf.org/html/rfc7845
+    // https://tools.ietf.org/html/rfc3533
+    class OpusFileStream : IDisposable {
```
Comment (vocoded) [inline examples/stream.cs:14]: please put opening brace for classes on a new line

File: examples/stream.cs [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+        int opusHeadersCount;
+        public int packetDurationMs { get; private set; }
+        public int framesPerPacket { get; private set; }
+        List<byte[]> savedPackets;
+
+        public OpusFileStream(string filename) {
+            try {
+                this.opusFile = File.OpenRead(filename);
+            } catch {
+                throw new Exception("Failed opening " + filename);
```
Comment (vocoded) [inline examples/stream.cs:31]: original exception will be lost, need to wrap

File: examples/stream.cs [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+            }
+            this.segmentSizes = new byte[255];
+            this.savedPackets = new List<byte[]>();
+            if (!this.fillOpusConfig()) {
+                throw new Exception("Failed to fill Opus configuration");    
+            }
+        }
+
+        public void Dispose() {
+            if (this.opusFile is null) {
```
Comment (vocoded) [inline examples/stream.cs:41]: please use `==` for null comparisons

File: examples/stream.cs [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+                        verifiedBytes++;
+                        if (verifiedBytes == 4) {
+                            return true;
+                        }
+                    } else {
+                        verifiedBytes = 0;
+                    }
+                }
+            } catch {
+                Console.WriteLine("Failed reading media while looking for next Ogg packet");
```
Comment (vocoded) [inline examples/stream.cs:65]: should we be swallowing this?  use of console.writeline should be avoided here

File: examples/stream.cs [hunk 6]
```diff
@@ -0,0 +1,638 @@
...
+                return false;
```
Comment (vocoded) [inline examples/stream.cs:66]: this return is redundant

File: examples/stream.cs [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+            // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+            // | ...                                                           | 28-
+            // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+            byte[] header = new byte[23];
+            try {
+                if (this.opusFile.Read(header, 0, header.Length) != header.Length) {
+                    return false;
+                }
+            } catch {
+                Console.WriteLine("Failed reading media while parsing Ogg packet header");
```
Comment (vocoded) [inline examples/stream.cs:98]: same as above

File: examples/stream.cs [hunk 8]
```diff
@@ -0,0 +1,638 @@
...
+            // Return the bool isContinueNeeded to accumulate such lacing data.
+            while (this.segmentIdx < this.segmentsCount) {
+                byte segmentSize = this.segmentSizes[this.segmentIdx];
+                byte[] segment = new byte[segmentSize];
+                try {
+                    if (this.opusFile.Read(segment, 0, segmentSize) != segmentSize) {
+                        break;
+                    }
+                } catch {
+                    Console.WriteLine("Failed reading media while fetching Ogg page segment");
```
Comment (vocoded) [inline examples/stream.cs:147]: same as above

File: examples/stream.cs [hunk 9]
```diff
@@ -0,0 +1,638 @@
...
+                isContinueNeeded = (segmentSize == 255);
+                this.segmentIdx++;
+                if (!isContinueNeeded) {
+                    break;
+                }
+            }
+            return (data, isContinueNeeded);
+        }
+
+        private bool parseOpusheadHeader(byte[] data) {
```
Comment (vocoded) [inline examples/stream.cs:164]:
```suggestion
        private bool parseOpusHeadHeader(byte[] data) {

```

File: examples/stream.cs [hunk 10]
```diff
@@ -0,0 +1,638 @@
...
+            UInt16 preskip = BitConverter.ToUInt16(data, 10);
+            this.sampleRate = BitConverter.ToUInt32(data, 12);
+            Console.WriteLine("Opus version = " + version);
+            Console.WriteLine("Channel count = " + channels);
+            Console.WriteLine("Pre-skip = " + preskip);
+            Console.WriteLine("Sample rate = " + this.sampleRate);
+            return true;
+        }
+
+        private bool parseOpustagsHeader(byte[] data) {
```
Comment (vocoded) [inline examples/stream.cs:204]:
```suggestion
        private bool parseOpusTagsHeader(byte[] data) {

```

File: examples/stream.py [hunk 11]
```diff
@@ -0,0 +1,446 @@
+import asyncio
+import base64
+import json
+import time
+import aiohttp
+import socket
+import configparser
+
+WS_ENDPOINT="wss://zello.io/ws"
```
Comment (vocoded) [inline examples/stream.py:9]: minor, whitespace around operator

File: examples/stream.cs (hunk 2 above)
Comment (ihorserba) [reply examples/stream.cs]: changed

File: examples/stream.cs (hunk 3 above)
Comment (ihorserba) [reply examples/stream.cs]: Added handling of the original exception

File: examples/stream.cs (hunk 4 above)
Comment (ihorserba) [reply examples/stream.cs]: changed

File: examples/stream.cs (hunk 5 above)
Comment (ihorserba) [reply examples/stream.cs]: replaced error message and return with throwing an exception everywhere.

File: examples/stream.cs (hunk 6 above)
Comment (ihorserba) [reply examples/stream.cs]: removed as changed this to throwing an exception

File: examples/stream.cs (hunk 7 above)
Comment (ihorserba) [reply examples/stream.cs]: changed

File: examples/stream.cs (hunk 8 above)
Comment (ihorserba) [reply examples/stream.cs]: changed

File: examples/stream.cs (hunk 9 above)
Comment (ihorserba) [reply examples/stream.cs]: replaced

File: examples/stream.cs (hunk 10 above)
Comment (ihorserba) [reply examples/stream.cs]: replaced

File: examples/stream.py (hunk 11 above)
Comment (ihorserba) [reply examples/stream.py]: added whitespaces

**Comment by ihorserba** [comment]
> I'll provide you references for style conventions offline

Changed curly brackets style and naming according the provided conventions

File: examples/README.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
+# Examples of using Zello Channel API
```
Comment (vocoded) [inline examples/README.md:1]:
```suggestion
# Examples of using the Zello Channel API
```

File: examples/README.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+There are example applications for the following platforms:
+
+* Python 3
+* .NET Core 3.1
+* Node.js 6
+
+Each application streams an audio file to Zello channel as regular voice message using common Zello account.
```
Comment (vocoded) [inline examples/README.md:9]:
```suggestion
Each application streams an audio file to a Zello channel as a regular voice message using a common Zello account.
```

File: examples/README.md [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+# Prepare a compatible audio source file
+
+The input format can be Wave, AIFF, FLAC, Ogg/FLAC, or raw PCM.
+
+Use appropriate converter for other media formats.
+
+
+## Encode the audio file using Opus codec
+
+Do not use frame size less than 20 milliseconds.
```
Comment (vocoded) [inline examples/README.md:41]:
```suggestion
Do not use a frame size less than 20 milliseconds.
```

File: examples/README.md [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+### Windows:
+
+Download the precompiled [binaries](https://archive.mozilla.org/pub/opus/win32/opus-tools-0.2-opus-1.3.zip) and unpack.
+
+
+# Prepare a compatible audio source file
+
+The input format can be Wave, AIFF, FLAC, Ogg/FLAC, or raw PCM.
+
+Use appropriate converter for other media formats.
```
Comment (vocoded) [inline examples/README.md:36]:
```suggestion
Use an appropriate converter for other media formats.
```

File: examples/README.md [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+
+
+# .NET example API client:
+```
+cd cs
+dotnet build
+dotnet run
+```
+
+# Applications architecture
```
Comment (vocoded) [inline examples/README.md:104]:
```suggestion
# Application architecture
```

File: examples/README.md [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+
+The `data` field in `stream data` defined in [Zello Channel API specification](../API.md) must be an audio encoded with [Opus codec](https://tools.ietf.org/html/rfc6716).
```
Comment (vocoded) [inline examples/README.md:106]:
```suggestion
The `data` field in `stream data` defined in the [Zello Channel API specification](../API.md) must be audio encoded with the [Opus codec](https://tools.ietf.org/html/rfc6716).
```

File: examples/cs/ConcentusAdapter.cs [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+            this.OpusFile.Dispose();
+        }
+
+        public OpusFileStream(string filename)
+        {
+            int samplesPerPacket;
+
+            this.FileName = filename;
+            this.OpusFile = new FileStream(filename, FileMode.Open);
+            this.Decoder = new OpusDecoder(48000, 1);
```
Comment (vocoded) [inline examples/cs/ConcentusAdapter.cs:34]: extract constant?

File: examples/cs/ConcentusAdapter.cs [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+            if (!this.OpusOggReadStream.HasNextPacket)
+            {
+                return null;
+            }
+
+            return this.OpusOggReadStream.RetrieveNextPacket();
+        }
+    }
+}
```
Comment (vocoded) [inline examples/cs/ConcentusAdapter.cs:67]: newline

File: examples/cs/IZelloOpusStream.cs [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+
+    interface IZelloOpusStream : IDisposable
+    {
+        string FileName { get; }
+        UInt32 SampleRate { get; }
+        int PacketDurationMs { get; }
+        int FramesPerPacket { get; }
+        byte[] GetNextOpusPacket();
+    }
+}
```
Comment (vocoded) [inline examples/cs/IZelloOpusStream.cs:13]: newline

File: examples/cs/OpusStream.cs [hunk 21]
```diff
@@ -0,0 +1,428 @@
...
+{
+    using System;
+    using System.IO;
+    using System.Collections.Generic;
+
+    // https://tools.ietf.org/html/rfc7845
+    // https://tools.ietf.org/html/rfc3533
+    class OpusFileStream : IZelloOpusStream
+    {
+	    Stream OpusFile;
```
Comment (vocoded) [inline examples/cs/OpusStream.cs:11]: indentation

File: examples/cs/OpusStream.cs [hunk 22]
```diff
@@ -0,0 +1,428 @@
...
+            this.SavedPackets = new List<byte[]>();
+            if (!this.FillOpusConfig())
+            {
+                throw new Exception("Failed to fill Opus configuration");
+            }
+        }
+
+        public void Dispose()
+        {
+            if (this.OpusFile == null)
```
Comment (vocoded) [inline examples/cs/OpusStream.cs:44]: as there's just a single call, let's invert this:

```
if (this.OpusFile != null) 
{
    this.OpusFile.Dispose();
}
```

**Review by vocoded — APPROVED** [review]

File: examples/cs/OpusStream.cs (hunk 22 above)
Comment (ihorserba) [reply examples/cs/OpusStream.cs]: Changed

File: examples/cs/OpusStream.cs (hunk 21 above)
Comment (ihorserba) [reply examples/cs/OpusStream.cs]: Fixed

File: examples/cs/IZelloOpusStream.cs (hunk 20 above)
Comment (ihorserba) [reply examples/cs/IZelloOpusStream.cs]: Added new line

File: examples/cs/ConcentusAdapter.cs (hunk 19 above)
Comment (ihorserba) [reply examples/cs/ConcentusAdapter.cs]: Added

File: examples/cs/ConcentusAdapter.cs (hunk 18 above)
Comment (ihorserba) [reply examples/cs/ConcentusAdapter.cs]: Added const as private class member

File: examples/README.md (hunk 16 above)
Comment (ihorserba) [reply examples/README.md]: Fixed

File: examples/README.md (hunk 15 above)
Comment (ihorserba) [reply examples/README.md]: Fixed

File: examples/README.md (hunk 14 above)
Comment (ihorserba) [reply examples/README.md]: Changed

File: examples/README.md (hunk 13 above)
Comment (ihorserba) [reply examples/README.md]: Changed

File: examples/README.md (hunk 12 above)
Comment (ihorserba) [reply examples/README.md]: Changed

File: examples/README.md (hunk 17 above)
Comment (ihorserba) [reply examples/README.md]: Fixed

**Pull request merged**